Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift

```diff
@@ -175,6 +181,51 @@ extension StripeCardReaderService: CardReaderService {
             }
         }
     }
+
+    public func checkForUpdate() -> Future<CardReaderSoftwareUpdate, Error> {
+        return Future() { promise in
+            Terminal.shared.checkForUpdate { [weak self] (softwareUpdate, error) in
+                guard let self = self else {
+                    promise(.failure(CardReaderServiceError.softwareUpdate()))
+                    return
+                }
+
+                if let error = error {
+                    let underlyingError = UnderlyingError(with: error)
+                    promise(.failure(CardReaderServiceError.softwareUpdate(underlyingError: underlyingError)))
+                }
+
+                if let softwareUpdate = softwareUpdate {
+                    self.pendingSoftwareUpdate = softwareUpdate
+                    let update = CardReaderSoftwareUpdate(update: softwareUpdate)
+                    promise(.success(update))
+                }
+            }
+        }
+    }
+
+    public func installUpdate() -> Future <Void, Error> {
+        return Future() { [weak self] promise in
+            guard let self = self,
+                  let pendingUpdate = self.pendingSoftwareUpdate else {
+                promise(.failure(CardReaderServiceError.softwareUpdate()))
+                return
+            }
+
+            // If the update succeeds the completion block is called with nil
+            // https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPTerminal.html#/c:objc(cs)SCPTerminal(im)installUpdate:delegate:completion:
+            Terminal.shared.installUpdate(pendingUpdate, delegate: self) { error in
+                if error == nil {
+                    promise(.success(()))
```

Comment:
Let's clear `self.pendingSoftwareUpdate` here too since it is no longer valid, right?

Reply:
Probably should clear it for success OR error

Reply:
I went with success only, the update might fail for several reasons, not clearing on error allows a retry.